File: litho-core/src/main/java/com/facebook/litho/DebugComponent.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.TextView;
import com.facebook.rendercore.MountDelegateTarget;
import com.facebook.rendercore.MountItem;
import com.facebook.rendercore.RenderTreeNode;
import com.facebook.rendercore.RenderUnit;
import com.facebook.rendercore.visibility.VisibilityMountExtension;
import com.facebook.rendercore.visibility.VisibilityOutput;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;

/**
 * A DebugComponent represents a node in Litho's component hierarchy. DebugComponent removes the
 * need to worry about implementation details of whether a node is represented by a {@link
 * Component} or a {@link ComponentLayout}. The purpose of this class is for tools such as Stetho's
 * UI inspector to be able to easily visualize a component hierarchy without worrying about
 * implementation details of Litho.
 */
public final class DebugComponent {

  public interface Overrider {
    void applyComponentOverrides(String key, Component component);

    void applyStateOverrides(String key, StateContainer state);

    void applyLayoutOverrides(String key, DebugLayoutNodeEditor node);
  }

  private static final Map<String, Overrider> sOverriders = new HashMap<>();

  private String mGlobalKey;
  private LithoLayoutResult mResult;
  private LithoNode mNode;
  private int mComponentIndex;
  private int mXOffset;
  private int mYOffset;
  private boolean mIsRoot;
  private @Nullable ComponentTreeTimeMachine mComponentTreeTimeMachine;

  private DebugComponent() {}

  static synchronized @Nullable DebugComponent getInstance(
      final LithoLayoutResult result,
      final int componentIndex,
      final int xOffset,
      final int yOffset,
      @Nullable final ComponentTree componentTree) {
    final DebugComponent debugComponent = new DebugComponent();
    final LithoNode node = result.getNode();
    final ComponentContext context = result.getContext();

    if (componentIndex >= node.getComponentCount()) {
      return null;
    }

    final String componentKey = node.getGlobalKeyAt(componentIndex);

    debugComponent.mComponentTreeTimeMachine =
        componentTree != null ? componentTree.getTimeMachine() : null;
    debugComponent.mGlobalKey = generateGlobalKey(context, componentKey);
    debugComponent.mResult = result;
    debugComponent.mNode = result.getNode();
    debugComponent.mComponentIndex = componentIndex;
    debugComponent.mXOffset = xOffset;
    debugComponent.mYOffset = yOffset;

    node.registerDebugComponent(debugComponent);

    return debugComponent;
  }

  @Nullable
  public static DebugComponent getRootInstance(LithoView view) {
    return getRootInstance(view.getComponentTree());
  }

  @Nullable
  public static DebugComponent getRootInstance(@Nullable ComponentTree componentTree) {
    final LayoutState layoutState =
        componentTree == null ? null : componentTree.getMainThreadLayoutState();
    final LithoLayoutResult root = layoutState == null ? null : layoutState.getRootLayoutResult();
    if (root == null) {
      return null;
    }
    final LithoNode node = root.getNode();
    final int outerWrapperComponentIndex = Math.max(0, node.getComponentCount() - 1);
    DebugComponent component =
        DebugComponent.getInstance(root, outerWrapperComponentIndex, 0, 0, componentTree);
    if (component != null) {
      component.mIsRoot = true;
    }
    return component;
  }

  @Nullable
  public static DebugComponent getInstance(LithoLayoutResult result) {
    final LithoNode rootNode = result.getNode();
    final int outerWrapperComponentIndex = Math.max(0, rootNode.getComponentCount() - 1);
    return DebugComponent.getInstance(result, outerWrapperComponentIndex, 0, 0, null);
  }

  public static @Nullable RenderUnit getRenderUnit(
      final DebugComponent debugComponent, final ComponentTree componentTree) {
    final Component component = debugComponent.getComponent();
    final LayoutState layoutState = componentTree.getMainThreadLayoutState();
    if (layoutState == null) {
      return null;
    }

    for (int i = 0, size = layoutState.getMountableOutputCount(); i < size; i++) {
      final RenderTreeNode renderTreeNode = layoutState.getMountableOutputAt(i);
      final LithoRenderUnit lithoRenderUnit = (LithoRenderUnit) renderTreeNode.getRenderUnit();
      if (lithoRenderUnit.componentContext != null
          && lithoRenderUnit.componentContext.getComponentScope() == component) {
        return lithoRenderUnit;
      }
    }

    return null;
  }

  public static @Nullable VisibilityOutput getVisibilityOutput(
      final DebugComponent debugComponent, final ComponentTree componentTree) {
    final String componentGlobalKey = debugComponent.getComponentGlobalKey();
    final LayoutState layoutState = componentTree.getMainThreadLayoutState();
    if (layoutState == null) {
      return null;
    }

    for (int i = 0, size = layoutState.getVisibilityOutputCount(); i < size; i++) {
      final VisibilityOutput visibilityOutput = layoutState.getVisibilityOutputAt(i);
      if (visibilityOutput.getId().equals(componentGlobalKey)) {
        return visibilityOutput;
      }
    }

    return null;
  }

  public static boolean isVisible(DebugComponent debugComponent, final LithoView lithoView) {
    final String componentGlobalKey = debugComponent.getComponentGlobalKey();
    final VisibilityMountExtension.VisibilityMountExtensionState visibilityState =
        lithoView.getVisibilityExtensionState();

    return VisibilityMountExtension.isVisible(visibilityState, componentGlobalKey);
  }

  private static String generateGlobalKey(ComponentContext context, String componentKey) {
    return System.identityHashCode(context.getLithoTree()) + componentKey;
  }

  static void applyOverrides(ComponentContext context, Component component, String componentKey) {
    final String key = generateGlobalKey(context, componentKey);
    final Overrider overrider = sOverriders.get(key);
    if (overrider != null) {
      overrider.applyComponentOverrides(key, component);
      overrider.applyStateOverrides(key, context.getScopedComponentInfo().getStateContainer());
    }
  }

  static void applyOverrides(ComponentContext context, LithoNode node) {
    if (node.getComponentCount() == 0) {
      return;
    }

    final String componentkey = node.getGlobalKeyAt(0);

    final String key = generateGlobalKey(context, componentkey);
    final Overrider overrider = sOverriders.get(key);
    if (overrider != null) {
      overrider.applyLayoutOverrides(key, new DebugLayoutNodeEditor(node));
    }
  }

  public void setOverrider(Overrider overrider) {
    sOverriders.put(mGlobalKey, overrider);
  }

  /**
   * Get the list of components composed by this component. This will not include any {@link View}s
   * that are mounted by this component as those are not components. Use {@link this#getMountedView}
   * for that.
   *
   * @return A list of child components.
   */
  public List<DebugComponent> getChildComponents() {

    if (isNotTailComponent()) {
      return getImmediateDescendantAsChild();
    } else {
      if (mResult instanceof NestedTreeHolderResult) {
        final LithoLayoutResult nestedResult = ((NestedTreeHolderResult) mResult).getNestedResult();

        if (nestedResult == null) {
          return Collections.emptyList();
        }

        if (nestedResult.mNode.getComponentCount() == 1) {
          if (nestedResult.getChildCount() == 0) {
            return Collections.emptyList();
          } else {
            getChildren(nestedResult, getXFromRoot(), getYFromRoot());
          }
        }

        final int index = Math.max(0, nestedResult.getNode().getComponentCount() - 2);
        final DebugComponent component =
            getInstance(nestedResult, index, getXFromRoot(), getYFromRoot(), null);
        return Collections.singletonList(component);

      } else {
        return getChildren(mResult, getXFromRoot(), getYFromRoot());
      }
    }
  }

  private boolean isNotTailComponent() {
    return mComponentIndex != 0;
  }

  private List<DebugComponent> getImmediateDescendantAsChild() {
    final int index = mComponentIndex - 1;
    if (index < 0) {
      return Collections.emptyList();
    }
    DebugComponent component = getInstance(mResult, index, mXOffset, mYOffset, null);
    return component != null ? Collections.singletonList(component) : Collections.emptyList();
  }

  private static List<DebugComponent> getChildren(LithoLayoutResult result, int x, int y) {
    final List<DebugComponent> children = new ArrayList<>();
    for (int i = 0, count = result.getChildCount(); i < count; i++) {
      final LithoLayoutResult childNode = result.getChildAt(i);
      final int index = Math.max(0, childNode.getNode().getComponentCount() - 1);
      DebugComponent component = getInstance(childNode, index, x, y, null);
      if (component != null) {
        children.add(component);
      }
    }
    return children;
  }

  /** @return A mounted view or null if this component does not mount a view. */
  @Nullable
  public View getMountedView() {
    final @Nullable Object content = getMountedContent();
    if (content instanceof View) {
      return (View) content;
    }

    return null;
  }

  /** @return A mounted drawable or null if this component does not mount a drawable. */
  @Nullable
  public Drawable getMountedDrawable() {
    final @Nullable Object content = getMountedContent();
    if (content instanceof Drawable) {
      return (Drawable) content;
    }

    return null;
  }

  /** @return The litho view hosting this component. */
  @Nullable
  public LithoView getLithoView() {
    final ComponentContext c = mResult.getContext();
    return c == null ? null : (LithoView) c.getMountedView();
  }

  /** @return The bounds of this component relative to its hosting {@link LithoView}. */
  public Rect getBoundsInLithoView() {
    if (isRoot()) {
      return new Rect(0, 0, mResult.getWidth(), mResult.getHeight());
    }

    final int x = getXFromRoot();
    final int y = getYFromRoot();
    return new Rect(x, y, x + mResult.getWidth(), y + mResult.getHeight());
  }

  /** @return The bounds of this component relative to its parent. */
  public Rect getBounds() {
    final int x = mResult.getX();
    final int y = mResult.getY();
    return new Rect(x, y, x + mResult.getWidth(), y + mResult.getHeight());
  }

  /**
   * The bounds of this component relative to its parent componen
   *
   * @return
   */
  public Rect getBoundsInParentDebugComponent() {

    /**
     * Many components can result in a single Lithonode/layout resut (from delegatation / custom
     * components) We want only the first component to 'take' the offset of the underlying layout
     * node other wise each delegating custom component will appear to multiply the offset.
     *
     * <p>NestedTreeHolder nodes have to be handled differently. The head component on the
     * NestedTreeHolderResult will create a separate DebugComponent node but any margin that needs
     * to be added to its bounds will be applied on the nested result node, which is hosted on
     * another DebugComponent instance.
     */
    boolean isHeadComponent = mComponentIndex == mNode.getComponentCount() - 1;
    final LithoLayoutResult nestedResult =
        (mResult instanceof NestedTreeHolderResult)
            ? ((NestedTreeHolderResult) mResult).getNestedResult()
            : null;
    final int xFromNestedResult = nestedResult == null ? 0 : nestedResult.getX();
    final int yFromNestedResult = nestedResult == null ? 0 : nestedResult.getY();
    final int x = mResult != null && isHeadComponent ? mResult.getX() + xFromNestedResult : 0;
    final int y = mResult != null && isHeadComponent ? mResult.getY() + yFromNestedResult : 0;

    return new Rect(x, y, x + mResult.getWidth(), y + mResult.getHeight());
  }

  /** @return True if this and given debug components share the same internal node */
  public boolean isSameNode(DebugComponent other) {
    return this.mNode == other.mNode;
  }

  /** @return the {@link ComponentContext} for this component. */
  public ComponentContext getContext() {
    return mResult.getContext();
  }

  /** @return True if this not has layout information attached to it (backed by a Yoga node) */
  public boolean isLayoutNode() {
    return mComponentIndex == 0;
  }

  /** @return This component's testKey or null if none is set. */
  @Nullable
  public String getTestKey() {
    return isLayoutNode() ? mNode.getTestKey() : null;
  }

  /**
   * Returns this component's testKey or null if none is set.
   *
   * <p>Unlike {@link #getTestKey()}, this function can return a test key set on any Component,
   * including container Components which resolve into LayoutNodes.
   *
   * <p>Unlike {@link #getTestKey()}, this function can also return test keys set on individual
   * Components even when they are all resolved into a single InternalNode.
   */
  @Nullable
  public String getComponentTestKey() {
    CommonProps props = mNode.getCommonPropsAt(mComponentIndex);
    return props == null ? null : props.getTestKey();
  }

  /**
   * @return This component's componentTag or null if none is set. Unlike {@link #getTestKey}, this
   *     will return tags for any Component, including Components which are not LayoutNodes.
   */
  @Nullable
  public Object getComponentTag() {
    CommonProps props = mNode.getCommonPropsAt(mComponentIndex);
    return props != null ? props.getComponentTag() : null;
  }

  /**
   * @return A concatenated string of all text content within the underlying LithoView. Null if the
   *     node doesn't have an associated LithoView.
   */
  @Nullable
  public String getAllTextContent() {
    final LithoView lithoView = getLithoView();

    if (lithoView == null) {
      return null;
    }

    final MountDelegateTarget mountDelegateTarget = lithoView.getMountDelegateTarget();
    final StringBuilder sb = new StringBuilder();

    for (int i = 0, size = mountDelegateTarget.getMountItemCount(); i < size; i++) {
      final MountItem mountItem = mountDelegateTarget.getMountItemAt(i);
      final Component mountItemComponent =
          mountItem == null ? null : LithoRenderUnit.getRenderUnit(mountItem).getComponent();
      if (mountItemComponent != null) {
        final Object content = mountItem.getContent();

        if (content instanceof TextContent) {
          for (CharSequence charSequence : ((TextContent) content).getTextList()) {
            sb.append(charSequence);
          }
        } else if (content instanceof TextView) {
          sb.append(((TextView) content).getText());
        }
      }
    }

    return sb.toString();
  }

  /**
   * @return The text content of the component wrapped by the debug component, or null if no
   *     TextContent/TextView are found.
   */
  @Nullable
  public String getTextContent() {
    final LithoView lithoView = getLithoView();
    if (lithoView == null) {
      return null;
    }

    final Component component = getComponent();
    final MountDelegateTarget mountDelegateTarget = lithoView.getMountDelegateTarget();
    for (int i = 0, size = mountDelegateTarget.getMountItemCount(); i < size; i++) {
      final MountItem mountItem = mountDelegateTarget.getMountItemAt(i);
      final Component mountItemComponent =
          mountItem == null ? null : LithoRenderUnit.getRenderUnit(mountItem).getComponent();
      if (mountItemComponent != null && mountItemComponent.getId() == component.getId()) {
        final Object content = mountItem.getContent();
        final StringBuilder sb = new StringBuilder();
        if (content instanceof TextContent) {
          for (CharSequence charSequence : ((TextContent) content).getTextList()) {
            sb.append(charSequence);
          }
        } else if (content instanceof TextView) {
          sb.append(((TextView) content).getText());
        }
        if (sb.length() != 0) {
          return sb.toString();
        }
      }
    }
    return null;
  }

  /** @return The {@link ComponentHost} that wraps this component or null if one cannot be found. */
  @Nullable
  public ComponentHost getComponentHost() {
    final LithoView lithoView = getLithoView();
    final Component component = getComponent();

    if (lithoView == null) {
      return null;
    }
    final MountDelegateTarget mountDelegateTarget = lithoView.getMountDelegateTarget();
    for (int i = 0, size = mountDelegateTarget.getMountItemCount(); i < size; i++) {
      final MountItem mountItem = mountDelegateTarget.getMountItemAt(i);
      final Component mountItemComponent =
          mountItem == null ? null : LithoRenderUnit.getRenderUnit(mountItem).getComponent();
      if (mountItemComponent != null && mountItemComponent.isEquivalentTo(component)) {
        return (ComponentHost) mountItem.getHost();
      }
    }

    return null;
  }

  /** @return This component's key or null if none is set. */
  public @Nullable String getKey() {
    final Component component = mNode.getComponentAt(mComponentIndex);
    return component.hasManualKey() ? component.getKey() : null;
  }

  /** @return The Component instance this debug component wraps. */
  public Component getComponent() {
    return mNode.getComponentAt(mComponentIndex);
  }

  @Nullable
  public ComponentTreeTimeMachine getComponentTreeTimeMachine() {
    return mComponentTreeTimeMachine;
  }

  private String getComponentGlobalKey() {
    return mNode.getComponentContextAt(mComponentIndex).getGlobalKey();
  }

  /** @return If this debug component represents a layout node, return it. */
  @Nullable
  public DebugLayoutNode getLayoutNode() {
    if (isLayoutNode()) {
      return new DebugLayoutNode(mResult);
    }
    return null;
  }

  public void rerender() {
    final LithoView lithoView = getLithoView();
    if (lithoView != null) {
      lithoView.forceRelayout();
    }
  }

  @Nullable
  public StateContainer getStateContainer() {
    return mNode.getComponentInfoAt(mComponentIndex).getStateContainer();
  }

  private int getXFromRoot() {
    if (mResult == null) {
      return 0;
    }
    return mResult.getX() + mXOffset;
  }

  private int getYFromRoot() {
    if (mResult == null) {
      return 0;
    }
    return mResult.getY() + mYOffset;
  }

  public String getGlobalKey() {
    return mGlobalKey;
  }

  public boolean canResolve() {
    return getComponent().canResolve();
  }

  public boolean isRoot() {
    return mComponentIndex == 0 && mIsRoot;
  }

  @Nullable
  private Object getMountedContent() {
    if (!isLayoutNode()) {
      return null;
    }

    final ComponentContext context = mResult.getContext();
    final LithoView view = context == null ? null : (LithoView) context.getMountedView();
    final MountDelegateTarget mountDelegateTarget =
        view == null ? null : view.getMountDelegateTarget();

    if (mountDelegateTarget != null) {
      for (int i = 0, count = mountDelegateTarget.getMountItemCount(); i < count; i++) {
        final MountItem mountItem = mountDelegateTarget.getMountItemAt(i);
        final Component component =
            mountItem == null ? null : LithoRenderUnit.getRenderUnit(mountItem).getComponent();

        if (component != null && component == mNode.getTailComponent()) {
          return mountItem.getContent();
        }
      }
    }

    return null;
  }
}

File: litho-core/src/main/java/com/facebook/litho/DebugComponent.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import android.widget.TextView
import com.facebook.litho.LithoRenderUnit.Companion.getRenderUnit
import com.facebook.rendercore.ClassBinderKey
import com.facebook.rendercore.RenderUnit
import com.facebook.rendercore.incrementalmount.ExcludeFromIncrementalMountBinder
import com.facebook.rendercore.visibility.VisibilityMountExtension
import com.facebook.rendercore.visibility.VisibilityOutput

/**
 * A DebugComponent represents a node in Litho's component hierarchy. DebugComponent removes the
 * need to worry about implementation details of whether a node is represented by a [ ] or a
 * [ComponentLayout]. The purpose of this class is for tools such as Flipper's UI inspector to be
 * able to easily visualize a component hierarchy without worrying about implementation details of
 * Litho.
 */
class DebugComponent
private constructor(
    val globalKey: String,
    private val result: LithoLayoutResult,
    private val node: LithoNode,
    private val componentIndex: Int,
    private val x: Int,
    private val y: Int,
    private val xOffset: Int,
    private val yOffset: Int,
) {

  interface Overrider {
    fun applyComponentOverrides(key: String, component: Component)

    fun applyStateOverrides(key: String, state: StateContainer)

    fun applyLayoutOverrides(key: String, node: DebugLayoutNodeEditor)
  }

  var isRoot = false
    get() = componentIndex == 0 && field
    private set

  /** @return the [ComponentContext] for this component. */
  val context: ComponentContext
    get() = result.context

  /** @return The litho view hosting this component. */
  val lithoView: BaseMountingView?
    get() = result.context?.mountedView as BaseMountingView?

  /** @return If this debug component represents a layout node, return it. */
  val layoutNode: DebugLayoutNode?
    get() = if (isLayoutNode) DebugLayoutNode(result) else null

  /** @return True if this not has layout information attached to it (backed by a Yoga node) */
  val isLayoutNode: Boolean
    get() = componentIndex == 0

  /** @return This component's testKey or null if none is set. */
  val testKey: String?
    get() = if (isLayoutNode) node.testKey else null

  /**
   * Returns this component's testKey or null if none is set.
   *
   * Unlike [testKey], this function can return a test key set on any Component, including container
   * Components which resolve into LayoutNodes.
   *
   * Unlike [testKey], this function can also return test keys set on individual Components even
   * when they are all resolved into a single InternalNode.
   */
  val componentTestKey: String?
    get() = node.getCommonPropsAt(componentIndex)?.testKey

  /**
   * @return This component's componentTag or null if none is set. Unlike [testKey], this will
   *   return tags for any Component, including Components which are not LayoutNodes.
   */
  val componentTag: Any?
    get() = node.getCommonPropsAt(componentIndex)?.componentTag

  /** @return This component's key or null if none is set. */
  val key: String?
    get() = component.run { if (hasManualKey()) key else null }

  /** @return The Component instance this debug component wraps. */
  val component: Component
    get() = node.getComponentAt(componentIndex)

  private val componentGlobalKey: String
    get() = node.getComponentContextAt(componentIndex).globalKey

  val stateContainer: StateContainer?
    get() = node.getComponentInfoAt(componentIndex).stateContainer

  private val xFromRoot: Int
    get() = x + xOffset

  private val yFromRoot: Int
    get() = y + yOffset

  private val isNotTailComponent: Boolean
    get() = componentIndex != 0

  val widthSpec: String
    get() = SizeSpec.toSimpleString(result.widthSpec)

  val heightSpec: String
    get() = SizeSpec.toSimpleString(result.heightSpec)

  fun setOverrider(overrider: Overrider) {
    overriders[globalKey] = overrider
  }

  /**
   * Get the list of components composed by this component. This will not include any [View]s that
   * are mounted by this component as those are not components. Use [this.getMountedView] for that.
   *
   * @return A list of child components.
   */
  val childComponents: List<DebugComponent>
    get() {
      if (result is NullLithoLayoutResult) {
        return emptyList()
      } else {
        return when {
          isNotTailComponent -> getImmediateDescendantAsChild()
          result is DeferredLithoLayoutResult -> {
            val nestedResult = result.result ?: return emptyList()
            if (nestedResult.node.componentCount == 1) {
              return when (nestedResult.childCount) {
                0 -> emptyList()
                else -> getChildren(nestedResult, xFromRoot, yFromRoot)
              }
            }
            val index = (nestedResult.node.componentCount - 2).coerceAtLeast(0)
            val component =
                getInstance(
                    nestedResult,
                    index,
                    result.getXForChildAtIndex(0),
                    result.getYForChildAtIndex(0),
                    xFromRoot,
                    yFromRoot,
                )
            listOfNotNull(component)
          }
          else -> getChildren(result, xFromRoot, yFromRoot)
        }
      }
    }

  private fun getImmediateDescendantAsChild(): List<DebugComponent> {
    val index = componentIndex - 1
    if (index < 0) {
      return emptyList()
    }
    val component = getInstance(result, index, x, y, xOffset, yOffset)
    return listOfNotNull(component)
  }

  /** @return A mounted view or null if this component does not mount a view. */
  val mountedView: View?
    get() = mountedContent as? View

  /** @return A mounted drawable or null if this component does not mount a drawable. */
  val mountedDrawable: Drawable?
    get() = mountedContent as? Drawable

  /** @return The bounds of this component relative to its hosting [LithoView]. */
  val boundsInLithoView: Rect
    get() {
      if (this.isRoot) {
        return Rect(0, 0, result.width, result.height)
      }
      val x = xFromRoot
      val y = yFromRoot
      return Rect(x, y, x + result.width, y + result.height)
    }

  /** @return The bounds of this component relative to its parent. */
  val bounds: Rect
    get() {
      return Rect(x, y, x + result.width, y + result.height)
    }

  /**
   * Many components can result in a single Lithonode/layout resut (from delegatation / custom
   * components) We want only the first component to 'take' the offset of the underlying layout node
   * other wise each delegating custom component will appear to multiply the offset.
   *
   * DeferredNode nodes have to be handled differently. The head component on the DeferredNodeResult
   * will create a separate DebugComponent node but any margin that needs to be added to its bounds
   * will be applied on the nested result node, which is hosted on another DebugComponent instance.
   */
  /**
   * The bounds of this component relative to its parent componen
   *
   * @return
   */
  val boundsInParentDebugComponent: Rect
    get() {
      /**
       * Many components can result in a single Lithonode/layout resut (from delegatation / custom
       * components) We want only the first component to 'take' the offset of the underlying layout
       * node other wise each delegating custom component will appear to multiply the offset.
       *
       * DeferredNode nodes have to be handled differently. The head component on the
       * DeferredNodeResult will create a separate DebugComponent node but any margin that needs to
       * be added to its bounds will be applied on the nested result node, which is hosted on
       * another DebugComponent instance.
       */
      val isHeadComponent = componentIndex == node.componentCount - 1
      val xFromNestedResult: Int
      val yFromNestedResult: Int
      if (result is DeferredLithoLayoutResult) {
        xFromNestedResult = result.getXForChildAtIndex(0)
        yFromNestedResult = result.getYForChildAtIndex(0)
      } else {
        xFromNestedResult = 0
        yFromNestedResult = 0
      }
      val x = if (isHeadComponent) x + xFromNestedResult else 0
      val y = if (isHeadComponent) y + yFromNestedResult else 0
      return Rect(x, y, x + result.width, y + result.height)
    }

  /** @return The alpha value of this component */
  val alpha: Float
    get() = result.node.nodeInfo?.alpha ?: 1.0f

  /** @return True if this and given debug components share the same internal node */
  fun isSameNode(other: DebugComponent): Boolean = node === other.node

  /**
   * @return A concatenated string of all text content within the underlying LithoView. Null if the
   *   node doesn't have an associated LithoView.
   */
  val allTextContent: String?
    get() = buildString {
      val mountDelegateTarget = lithoView?.mountDelegateTarget ?: return null
      for (i in 0 until mountDelegateTarget.getMountItemCount()) {
        val mountItem = mountDelegateTarget.getMountItemAt(i)
        val mountItemComponent = mountItem?.let { getRenderUnit(it).component }
        if (mountItemComponent != null) {
          val content = mountItem.content
          if (content is TextContent) {
            for (charSequence in content.textList) append(charSequence)
          } else if (content is TextView) {
            append(content.text)
          }
        }
      }
    }

  /**
   * @return The text content of the component wrapped by the debug component, or null if no
   *   TextContent/TextView are found.
   */
  val textContent: String?
    get() {
      val mountDelegateTarget = lithoView?.mountDelegateTarget ?: return null
      for (i in 0 until mountDelegateTarget.getMountItemCount()) {
        val mountItem = mountDelegateTarget.getMountItemAt(i)
        val mountItemComponent = mountItem?.let { getRenderUnit(it).component }
        if (mountItemComponent?.instanceId == component.instanceId) {
          val content = mountItem.content
          val sb = StringBuilder()
          if (content is TextContent) {
            for (charSequence in content.textList) sb.append(charSequence)
          } else if (content is TextView) {
            sb.append(content.text)
          }
          if (sb.isNotEmpty()) {
            return sb.toString()
          }
        }
      }
      return null
    }

  /** @return The [ComponentHost] that wraps this component or null if one cannot be found. */
  val componentHost: ComponentHost?
    get() {
      val mountDelegateTarget = lithoView?.mountDelegateTarget ?: return null
      for (i in 0 until mountDelegateTarget.getMountItemCount()) {
        val mountItem = mountDelegateTarget.getMountItemAt(i)
        val mountItemComponent = mountItem?.let { getRenderUnit(it).component }
        if (mountItemComponent?.isEquivalentTo(component) == true) {
          return mountItem.host as ComponentHost?
        }
      }
      return null
    }

  fun rerender() {
    (lithoView as? LithoView)?.forceRelayout()
  }

  fun canResolve(): Boolean {
    return component is SpecGeneratedComponent &&
        TestLayoutState.canResolve(component as SpecGeneratedComponent)
  }

  val mountedContent: Any?
    get() {
      if (!isLayoutNode) {
        return null
      }
      val mountDelegateTarget = lithoView?.mountDelegateTarget
      if (mountDelegateTarget != null) {
        for (i in 0 until mountDelegateTarget.getMountItemCount()) {
          val mountItem = mountDelegateTarget.getMountItemAt(i)
          val component = mountItem?.let { getRenderUnit(it).component }
          if (component != null && component === node.tailComponent) {
            return mountItem.content
          }
        }
      }
      return null
    }

  companion object {
    private val overriders: MutableMap<String, Overrider> = HashMap()

    @JvmStatic
    @Synchronized
    fun getInstance(
        result: LithoLayoutResult,
        componentIndex: Int,
        x: Int,
        y: Int,
        xOffset: Int,
        yOffset: Int,
    ): DebugComponent? {
      val node = result.node
      val context = result.context
      if (result is NullLithoLayoutResult || componentIndex >= node.componentCount) {
        return null
      }
      val componentKey = node.getGlobalKeyAt(componentIndex)
      return DebugComponent(
          globalKey = generateGlobalKey(context, componentKey),
          result = result,
          node = result.node,
          componentIndex = componentIndex,
          x = x,
          y = y,
          xOffset = xOffset,
          yOffset = yOffset,
      )
    }

    @JvmStatic
    fun getRootInstance(view: BaseMountingView): DebugComponent? =
        getRootInstance(view.currentLayoutState)

    fun getRootInstance(componentTree: ComponentTree?): DebugComponent? =
        if (componentTree == null) null else getRootInstance(componentTree.mainThreadLayoutState)

    @JvmStatic
    fun getRootInstance(
        layoutState: LayoutState?,
    ): DebugComponent? {
      val root = layoutState?.rootLayoutResult
      if (root == null || root is NullLithoLayoutResult) {
        return null
      }
      check(root is LithoLayoutResult) { "Expected root to be a LithoLayoutResult" }
      val node = root.node
      val outerWrapperComponentIndex = (node.componentCount - 1).coerceAtLeast(0)
      return getInstance(root, outerWrapperComponentIndex, 0, 0, 0, 0)?.apply { isRoot = true }
    }

    @JvmStatic
    fun getInstance(result: LithoLayoutResult, x: Int, y: Int): DebugComponent? {
      val rootNode = result.node
      val outerWrapperComponentIndex = (rootNode.componentCount - 1).coerceAtLeast(0)
      return getInstance(result, outerWrapperComponentIndex, x, y, 0, 0)
    }

    /**
     * Returns whether this Component should be considered to be mounted, even outside of
     * incremental mount.
     *
     * @see SpecGeneratedComponent.excludeFromIncrementalMount
     */
    @JvmStatic
    fun isExcludedFromIncrementalMount(
        debugComponent: DebugComponent,
        layoutState: LayoutState
    ): Boolean {
      val renderUnit = getRenderUnit(debugComponent, layoutState)
      val component = debugComponent.component

      val shouldExcludePrimitiveFromIncrementalMount: Boolean =
          renderUnit?.findAttachBinderByKey<ExcludeFromIncrementalMountBinder>(
              ClassBinderKey(ExcludeFromIncrementalMountBinder::class.java)) != null
      val shouldExcludeSpecGeneratedComponentFromIncrementalMount: Boolean =
          component is SpecGeneratedComponent && component.excludeFromIncrementalMount()

      return shouldExcludePrimitiveFromIncrementalMount ||
          shouldExcludeSpecGeneratedComponentFromIncrementalMount
    }

    @JvmStatic
    fun getRenderUnit(debugComponent: DebugComponent, layoutState: LayoutState): RenderUnit<*>? {
      val component = debugComponent.component
      for (i in 0 until layoutState.getMountableOutputCount()) {
        val renderTreeNode = layoutState.getMountableOutputAt(i)
        val lithoRenderUnit = renderTreeNode.renderUnit as LithoRenderUnit
        if (lithoRenderUnit.componentContext?.componentScope === component) {
          return lithoRenderUnit
        }
      }
      return null
    }

    @JvmStatic
    fun getVisibilityOutput(
        debugComponent: DebugComponent,
        layoutState: LayoutState
    ): VisibilityOutput? {
      val componentGlobalKey = debugComponent.componentGlobalKey
      for (i in 0 until layoutState.visibilityOutputCount) {
        val visibilityOutput = layoutState.getVisibilityOutputAt(i)
        if (visibilityOutput.id == componentGlobalKey) {
          return visibilityOutput
        }
      }
      return null
    }

    @JvmStatic
    fun isVisible(debugComponent: DebugComponent, lithoView: BaseMountingView): Boolean {
      val componentGlobalKey = debugComponent.componentGlobalKey
      return lithoView.visibilityExtensionState?.let {
        VisibilityMountExtension.isVisible(it, componentGlobalKey)
      } ?: false
    }

    private fun generateGlobalKey(context: ComponentContext, componentKey: String?): String =
        generateGlobalKey(context.lithoTree?.id, componentKey)

    @JvmStatic
    fun generateGlobalKey(treeId: Int?, globalKey: String?): String =
        "${treeId?: "notree"}:${globalKey}"

    @JvmStatic
    fun applyOverrides(context: ComponentContext, component: Component, componentKey: String) {
      val key = generateGlobalKey(context, componentKey)
      val overrider = overriders[key]
      if (overrider != null) {
        overrider.applyComponentOverrides(key, component)
        val stateContainer = context.scopedComponentInfo.stateContainer
        if (stateContainer != null) {
          overrider.applyStateOverrides(key, stateContainer)
        }
      }
    }

    @JvmStatic
    fun applyOverrides(context: ComponentContext, node: LithoNode) {
      if (node.componentCount == 0) {
        return
      }
      val componentKey = node.getGlobalKeyAt(0)
      val key = generateGlobalKey(context, componentKey)
      val overrider = overriders[key]
      overrider?.applyLayoutOverrides(key, DebugLayoutNodeEditor(node))
    }

    private fun getChildren(result: LithoLayoutResult, xOffset: Int, yOffset: Int) = buildList {
      for (i in 0 until result.childCount) {
        val childNode = result.getChildAt(i)
        val index = (childNode.node.componentCount - 1).coerceAtLeast(0)
        getInstance(
                childNode,
                index,
                result.getXForChildAtIndex(i),
                result.getYForChildAtIndex(i),
                xOffset,
                yOffset,
            )
            ?.let { add(it) }
      }
    }
  }
}
